Some lemmas to do with transposition lists

Taneb opened the pull request:
This is largely preliminaries for #1741, which I'm still playing with how to implement.

MatthewDaggitt commented on src/Data/Fin/Permutation/Transposition/List.agda:
```diff
@@ -29,43 +31,68 @@ private
 
 TranspositionList : ℕ → Set
-TranspositionList n = List (Fin n × Fin n)
+TranspositionList n = List (Σ[ i ∈ Fin n ] ∃[ j ] i ≢ j)
```
Maybe `List (∃₂ λ i j → i ≢ j)`? Could also write `(∃₂ _≢_)` but I'm not sure how clear that is?

Taneb replied:
You need a hint that they have type `Fin n` rather than some other type, sadly.

MatthewDaggitt replied:
`List (∃₂ λ (i j : Fin n) → i ≢ j)` then?

MatthewDaggitt commented on src/Data/Fin/Permutation/Transposition/List.agda:
```diff
@@ -29,43 +31,68 @@ private
 
 lift₀ : TranspositionList n → TranspositionList (suc n)
-lift₀ xs = map (λ (i , j) → (suc i , suc j)) xs
+lift₀ xs = map (λ (i , j , i≢j) → (suc i , suc j , λ si≡sj → i≢j (suc-injective si≡sj))) xs
```
Maybe `i≢j ∘ suc-injective` instead?

MatthewDaggitt commented on src/Data/Fin/Permutation.agda:
```diff
@@ -223,6 +223,13 @@ insert {m} {n} i j π = permutation to from inverseˡ′ inverseʳ′
 ------------------------------------------------------------------------
 -- Other properties
 
+transpose-self-inverse : ∀ (i j : Fin n) → transpose i j ≈ transpose j i
```
Also it's not in the CHANGELOG

Taneb replied:
It's more of a commutativity property than a symmetry property (the type is the same as `Commutative _≈_ transpose` I think, modulo unresolved metas). Would you be OK with `transpose-comm`?

jamesmckinna replied:
Since #1914 why not call this `transpose-selfInverse`?

MatthewDaggitt commented on CHANGELOG.md:
```diff
@@ -491,6 +491,9 @@ Non-backwards compatible changes
 * In `Data.Fin.Properties` the `i` argument to `opposite-suc` has been made explicit.
 
+* In `Data.Fin.Permutation.Transposition.List`, the transpositions are now restricted
+  to being pairs of distinct elements.
```
Could we also give the type before and the type after here?

MatthewDaggitt commented:
Apologies for taking so long for a review!

Taneb commented:
I started addressing the comments, but I've been very distracted lately (and have broken my wrist in the meantime). I don't know when I can get back to it

MatthewDaggitt commented:
Oh no, sorry to hear that! I can see why that'd make typing a little difficulty. Okay no worries, I'll leave it up and see if I have the time to fiddle with it in the meantime....

MatthewDaggitt commented on src/Data/Fin/Permutation/Transposition/List.agda:
```diff
@@ -29,43 +31,68 @@ private
 
 TranspositionList : ℕ → Set
-TranspositionList n = List (Fin n × Fin n)
+TranspositionList n = List (∃₂ λ (i j : Fin n) → i ≢ j)
```
Hmm I'm sorry to reopen the discussion, but looking at this with a fresh pair of eyes, wouldn't a more modular and reusable solution be to keep the definition as is and then define a predicate:
```
NonTrivialEntries : TranspositionList n → Set
NonTrivialEntries xs = All (curry _≢_) xs
```
which could then be taken as an argument in the proofs that need it. The advantages of this would be:

1. Preserves backwards compatibility
2. Provides more choice to the user - I can imagine a scenario where the reflexive entries do hold semantic information.
3. Allows us to reuse lemmas from `Data.List.Relation.Unary.All`

I'm open to suggestions about the name `NonTrivialEntries`?

Taneb replied:
I really want to argue against this, but the truth is that for everything currently existing, that the transposition list has no trivial entries is completely irrelevant. I think the right thing to do for now might be to not talk about that constraint until we've got something that actually uses it.

MatthewDaggitt replied:
Yup, happy with that solution!

JacquesCarette replied:
So what needs to be done to implement that solution?

Taneb commented:
Now that #1852 is in, I'd like to continue work on this PR to where I wanted to get it to, proving that there's a group homomorphism from permutations to their parity (this will let us define alternating groups and is useful for implementing Liebniz's formula for matrix determinants)

JacquesCarette commented on src/Data/Fin/Permutation/Components.agda:
```diff
@@ -28,17 +28,32 @@ open ≡-Reasoning
 
 -- 'tranpose i j' swaps the places of 'i' and 'j'.
 
+{-
```
Remove this commented-out code before actually merging.

JacquesCarette commented on src/Data/Fin/Permutation/Components.agda:
```diff
@@ -28,17 +28,32 @@ open ≡-Reasoning
 --  Properties
 --------------------------------------------------------------------------------
 
+transpose-comm : ∀ {n} (i j : Fin  n) → transpose i j ≗ transpose j i
```
This is now both here and in `Data.Fin.Permutation`. I think here is a better place for it.

JacquesCarette commented on src/Data/Fin/Permutation/Transposition/List.agda:
```diff
@@ -29,43 +38,208 @@ private
+        prop : transpose i j ∘ₚ transpose iₕ jₕ ∘ₚ eval xs ≈ eval xs
+        prop k rewrite iₕ≡j rewrite jₕ≡i = cong (eval xs ⟨$⟩ʳ_) (PC.transpose-inverse j i)
+  
+{-
```
delete commented out code

JacquesCarette commented on src/Data/List/Properties.agda:
```diff
@@ -75,6 +75,9 @@ module _ {x y : A} {xs ys : List A} where
 ≡-dec _≟_ []       (y ∷ ys) = no λ()
 ≡-dec _≟_ (x ∷ xs) (y ∷ ys) = ∷-dec (x ≟ y) (≡-dec _≟_ xs ys)
 
+length[xs]≡0⇒xs≡[] : {xs : List A} → length xs ≡ 0 → xs ≡ []
```
Personally I would decouple this simple property from this PR and make it a separate tiny PR.

MatthewDaggitt commented:
@Taneb I'm afraid in the end I didn't get a chance to look at it, and no real prospect of it happening soon. Any likelihood of you picking this backup in the next month? Otherwise I propose to close.

Taneb commented:
It's always been on my to-do list, but I want it to include the parity theorem. Feel free to close it and I'll open a new PR when it's ready

jamesmckinna commented:
Or you could rebase this PR and then subsequently re-open once you have Parity working...?
(sorry we didn't manage to push it over the line last year in Delft!)
